improver/psychrometric_calculations/condensation_trails.py

```diff
@@ -97,6 +103,175 @@ def _find_local_vapour_pressure(self, pressure_levels: np.ndarray) -> np.ndarray
         )
         return self.relative_humidity * svp
 
+    @staticmethod
+    def _critical_temperatures_and_intercepts_for_given_contrail_factor(
+        mixing_ratios: np.ndarray,
+        relative_humidity: np.ndarray,
+        svp: np.ndarray,
+        svp_derivative: np.ndarray,
+        temperature: np.ndarray,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Calculate the critical temperatures and critical intercepts on pressure levels
+        for a single engine contrail factor.
+
+        These are calculated at each pressure level by drawing a tangent to the saturation
+        vapour pressure curve with respect to water. The tangent gradient is equal to the
+        engine mixing ratio.
+
+        .. include:: extended_documentation/psychrometric_calculations/condensation_trails/critical_temperatures.rst
+
+        Args:
+            mixing_ratios (np.ndarray): Engine mixing ratios on pressure levels. Pressure is the leading axis (Pa/K).
+            relative_humidity (np.ndarray): Relative humidity on pressure levels. Pressure is the leading axis (kg/kg).
+            svp (np.ndarray): Lookup table of saturation vapour pressure with respect to water (Pa).
+            svp_derivative (np.ndarray): Lookup table of the first derivative of saturation vapour pressure with respect to water (Pa/K).
+            temperature (np.ndarray): Air temperatures corresponding to both lookup tables (K).
+
+        Returns:
+            Tuple[np.ndarray, np.ndarray]: Arrays of critical temperatures on pressure levels (K), and critical intercepts on pressure levels (Pa).
+
+        """
+        num_pressure_levels = mixing_ratios.shape[0]
+
+        # maximum critical temperature (at 100% relative humidity) is given by
+        # the point on the SVP derivative curve that is equal to the engine mixing
+        # ratio
+        ind_max = [(np.abs(svp_derivative - m)).argmin() for m in mixing_ratios]
+        tangent_gradient = svp_derivative[ind_max]
+        critical_temperature_maximum = temperature[ind_max]
+
+        # tangent to SVP curve, with gradient equal to the engine mixing ratio
+        critical_intercepts = (
+            svp[ind_max] - tangent_gradient * critical_temperature_maximum
+        )
+        tangent_vapour_pressure = (
+            tangent_gradient[:, np.newaxis] * temperature[np.newaxis, :]
+            + critical_intercepts[:, np.newaxis]
+        )
+
+        # minimum critical temperature (at 0% relative humidity) is given by
+        # point at which the tangent crosses the line of zero vapour pressure
+        ind_min = [
+            np.abs(tangent_vapour_pressure[i, :]).argmin()
+            for i in range(num_pressure_levels)
+        ]
```

Comment:
This doesn't look right to me. I was expecting to calculate the critical intercepts for each pressure level independently, so there shouldn't be anything scanning other pressure level points.

I think the issue is that air_pressure has been confused with vapour_pressure.